lapce-core/src/syntax/mod.rs

```diff
@@ -68,6 +69,300 @@ pub enum Error {
+            children: vec![],
+            level: 0,
+        }
+    }
+
+    pub fn new_with_type(tt: NodeType, len: usize) -> Self {
+        Self {
+            tt,
+            len,
+            children: vec![],
+            level: 0,
+        }
+    }
+}
+
+impl Default for ASTNode {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BracketParser {
+    pub code: Vec<char>,
+    pub cur: usize,
+    pub ast: ASTNode,
+    bracket_set: HashMap<char, ASTNode>,
+    pub bracket_pos: HashMap<usize, Vec<LineStyle>>,
+    mode: BracketParserMode,
+    noparsing_token: Vec<char>,
+    pub active: bool,
+}
+
+impl BracketParser {
+    pub fn new(code: String, active: bool) -> Self {
+        Self {
+            code: code.chars().collect(),
+            cur: 0,
+            ast: ASTNode::new(),
+            bracket_set: HashMap::from([
+                ('(', ASTNode::new_with_type(NodeType::LeftParen, 1)),
+                (')', ASTNode::new_with_type(NodeType::RightParen, 1)),
+                ('{', ASTNode::new_with_type(NodeType::LeftCurly, 1)),
+                ('}', ASTNode::new_with_type(NodeType::RightCurly, 1)),
+                ('[', ASTNode::new_with_type(NodeType::LeftBracket, 1)),
+                (']', ASTNode::new_with_type(NodeType::RightBracket, 1)),
+            ]),
+            bracket_pos: HashMap::new(),
+            mode: BracketParserMode::Parsing,
+            noparsing_token: vec!['\'', '"', '`'],
+            active,
+        }
+    }
+
+    /*pub fn enable(&self) {
+        *(self.active.borrow_mut()) = true;
+    }
+
+    pub fn disable(&self) {
+        *(self.active.borrow_mut()) = false;
+    }*/
+
+    pub fn update_code(
+        &mut self,
+        code: String,
+        buffer: &Buffer,
+        syntax: Option<Syntax>,
+    ) {
+        let palette = vec![
+            "bracket.color.1".to_string(),
+            "bracket.color.2".to_string(),
+            "bracket.color.3".to_string(),
+        ];
+        if self.active {
+            self.bracket_pos = HashMap::new();
+            if let Some(syntax) = syntax {
+                if let Some(tree) = syntax.layers.try_tree() {
+                    let mut walk_cursor = tree.walk();
+                    let mut bracket_pos: HashMap<usize, Vec<LineStyle>> =
+                        HashMap::new();
+                    language::walk_tree_bracket_ast(
+                        &mut walk_cursor,
+                        &mut 0,
+                        &mut 0,
+                        &mut bracket_pos,
+                        &palette,
+                    );
+                    self.bracket_pos = bracket_pos;
+                }
+            } else {
+                self.code = code.chars().collect();
+                self.cur = 0;
+                self.parse();
+                let mut pos_vec = vec![];
+                Self::highlight_pos(
+                    &self.ast,
+                    &mut pos_vec,
```

Comment:
Why pass a vec instead of just returning it?

Reply:
Since this is a recursive parser. Passing a vec can save my time to fight against the borrow checker. Trust me I tried to return when I wrote another similar parser and it was a nightmare.